lang_qc/db/helper/wells.py

```diff
@@ -133,18 +161,86 @@ def recent_completed_wells(self) -> List[PacBioRunWellMetrics]:
+    This class creates `PacBioPagedWellsLite` objects that correspond to
     the criteria given by the attributes of the object, i.e. `page_size`
-    `page_number` and `qc_flow_status` attributes.
+    and `page_number`.
     """
 
     qcdb_session: Session = Field(
         title="SQLAlchemy Session",
         description="A SQLAlchemy Session for the LangQC database",
     )
 
+    class Config:
+        arbitrary_types_allowed = True
+        extra = Extra.forbid
+        allow_mutation = True
+
+    def generate_well_models(
+        self, db_wells_list: List[PacBioRunWellMetrics], qc_flow_status: str = ""
+    ) -> List[PacBioWell]:
+        """
+        Given a list of well db records, generates a list of corresponding
+        PacBioWell objects, which are enhanced with the QC state information
+        if it is available.
+        """
+
+        # Normally QC data is not available for the inbox, aborted, etc.
+        # wells. If some well with a non-inbox status has QC state assigned,
+        # the same well will also be retrieved by the 'in progress' or
+        # 'on hold' or 'qc complete' queries. However, it is useful to display
+        # the QC state if it is available.
+
+        pb_wells = []
+        for db_well in db_wells_list:
+            attrs = {"run_name": db_well.pac_bio_run_name, "label": db_well.well_label}
+            if qc_flow_status != QcFlowStatusEnum.INBOX:
+                qc_state = WellQc(
+                    session=self.qcdb_session,
+                    run_name=db_well.pac_bio_run_name,
+                    well_label=db_well.well_label,
+                ).current_qc_state()
+                if qc_state:
+                    attrs["qc_state"] = qc_state
+            pb_well = PacBioWell.parse_obj(attrs)
+            pb_well.copy_run_tracking_info(db_well)
+            pb_wells.append(pb_well)
+
+        return pb_wells
+
+    def create_for_runs(self, run_names: List[str]) -> PacBioPagedWellsLite:
+        """
+        Returns `PacBioPagedWellsLite` object that corresponds to the criteria
+        specified by the `page_size` and `page_number` attributes.
+
+        The `PacBioWell` objects in `wells` attribute of the returned object
+        belong to runs specified in the `run_names` list argument and are sorted
+        by the run name and well label.
+        """
+
+        if len(run_names) == 0:
+            raise EmptyListOfRunNamesError("List of run names cannot be empty.")
+
+        wells = self.get_wells_in_runs(run_names)
+        total_number_of_wells = len(wells)
+        wells = self.slice_data(wells)
+        return PacBioPagedWellsLite(
+            page_number=self.page_number,
+            page_size=self.page_size,
+            total_number_of_items=total_number_of_wells,
+            wells=self.generate_well_models(wells),
+        )
+
+
+class PacBioPagedWellsFactory(PacBioPagedWellsFactoryLite, QcFlowStatusMixin):
```

Comment:
This is not very persuasive as a factory, because the create method is not accepting inputs. The inherited `create_for_runs` is factory-like but it should be used by this factory due to the nominatively excluded return type. I could see you having a factory for wells in general that emits the correct types by their implementations. That would be more gang-of-foury, but otherwise the inheritance and behaviours are getting rather muddy.